src/astra_rl/training/trainer.py

```diff
@@ -0,0 +1,163 @@
+        ...     optimizer="adamw",
+        ...     gradient_accumulation_steps=1,
+        ...     training_steps=1024,
+        ...     num_episodes_per_experience=8,
+        ... )
+        >>> trainer = Trainer(
+        ...     config,
+        ...     environment,
+        ...     algorithm,
+        ... )
+        >>> trainer.train()
+
+    Attributes:
+        config (TrainingConfiguration): The configuration for the training process.
+        harness (Harness): The harness that manages the training loop and interactions with the environment. See `astra_rl.training.harness` for what it does.
+        optimizer (Optimizer): The optimizer used for updating the model parameters.
+        _global_step_counter (int): A counter for global steps, used for gradient accumulation.
+    """
+
+    optimizer: Optimizer
+
+    def __init__(
+        self,
+        config: TrainingConfiguration,
+        environment: Environment[StateT, ActionT],
+        algorithm: Algorithm[StateT, ActionT, Step, Batch],
+    ):
+        """
+        Args:
+            config (TrainingConfiguration): The configuration for the training process.
+            environment (Environment): The environment to run our algorithm in.
+            algorithm (Algorithm): The algorithm used for training the attacker agent.
+        """
+
+        self.config = config
+        self.harness = Harness(
+            environment, algorithm, config.num_episodes_per_experience
+        )
+
+        # TODO initialize LR scheduler?
+        # ?????????????????????????????
+
+        # initialize optimizer
+        if config.optimizer == "adam":
+            from torch.optim import Adam
+
+            self.optimizer = Adam(environment.problem.parameters(), config.lr)
+        elif config.optimizer == "adamw":
+            from torch.optim import AdamW
+
+            self.optimizer = AdamW(environment.problem.parameters(), config.lr)
+        elif config.optimizer == "sgd":
+            from torch.optim import SGD
+
+            self.optimizer = SGD(environment.problem.parameters(), config.lr)
+        elif config.optimizer == "rmsprop":
+            from torch.optim import RMSprop
+
+            self.optimizer = RMSprop(environment.problem.parameters(), config.lr)
+        elif config.optimizer == "adagrad":
+            from torch.optim import Adagrad
+
+            self.optimizer = Adagrad(environment.problem.parameters(), config.lr)
+        else:
+            raise ValueError(f"Unknown optimizer configured: {config.optimizer}")
+
+        # step counter, for acccmulutaion, etc.
+        self._global_step_counter = 0
+
+    def train(self) -> None:
+        """Run training by the specified config!
+
+        Note:
+            This method takes no arguments and returns nothing, and its
+            only used for side effects. We don't really need it other than
+            its helpful for delaying the training process until we are ready
+            to run it for other initialization.
```

Comment:
This note really doesn't make sense to me. It doesn't seem to delay the training process, it appears to actually just execute the training process.

Also nit: "its helpful" should be "it's helpful" since the usage is a contraction of "it is helpful"

Reply:
Updated to

>             This method takes no arguments and returns nothing, and its
>            only used for side effects. We don't really need it other than
>            it's helpful for allowing the user to control when training
>            actually starts (instead of immediately after Trainer construction).